rust/stackablectl/src/cmds/debug.rs

```diff
@@ -0,0 +1,380 @@
+    task::{ready, Poll},
+};
+
+use clap::Args;
+use futures::{
+    channel::mpsc::{self, Sender},
+    FutureExt, SinkExt, TryFutureExt,
+};
+use rand::Rng;
+use snafu::{futures::TryFutureExt as _, OptionExt, ResultExt, Snafu};
+use stackable_operator::{
+    k8s_openapi::api::core::v1::{ContainerStatus, EphemeralContainer, Pod, PodSpec},
+    kube::{
+        self,
+        api::{AttachParams, PatchParams, TerminalSize},
+        runtime::reflector::ObjectRef,
+    },
+};
+use termion::{raw::IntoRawMode, terminal_size};
+use tokio::{
+    io::{unix::AsyncFd, AsyncRead},
+    signal::unix::SignalKind,
+};
+use tracing::{error, info, info_span, warn, Instrument};
+
+use crate::cli::Cli;
+
+#[derive(Debug, Snafu)]
+pub enum CmdError {
+    #[snafu(display("failed to create Kubernetes client"))]
+    KubeClientCreate { source: kube::Error },
+    #[snafu(display("failed to get {pod}"))]
+    GetPod {
+        source: kube::Error,
+        pod: ObjectRef<Pod>,
+    },
+    #[snafu(display("{pod} has no container {container:?}"))]
+    FindTemplateContainer {
+        pod: ObjectRef<Pod>,
+        container: String,
+    },
+    #[snafu(display("failed to create ephemeral debug container {container:?} on {pod}"))]
+    CreateDebugContainer {
+        source: kube::Error,
+        pod: ObjectRef<Pod>,
+        container: String,
+    },
+    #[snafu(display("debug container {container:?} on {pod} never became ready"))]
+    AwaitDebugContainerReadiness {
+        source: kube::runtime::wait::Error,
+        pod: ObjectRef<Pod>,
+        container: String,
+    },
+    #[snafu(display("failed to get status of debug container {container:?} on {pod}"))]
+    FindDebugContainerStatus {
+        pod: ObjectRef<Pod>,
+        container: String,
+    },
+    #[snafu(display("failed to attach to container {container:?} on {pod}"))]
+    AttachContainer {
+        source: kube::Error,
+        pod: ObjectRef<Pod>,
+        container: String,
+    },
+    #[snafu(display("failed to enable raw local TTY input"))]
+    SetRawTtyMode { source: std::io::Error },
+    #[snafu(display("failed to turn stdin async"))]
+    AsyncifyStdin { source: std::io::Error },
+    #[snafu(display("failed to initialize AsyncFd for stdin"))]
+    AsyncFdStdin { source: std::io::Error },
+    #[snafu(display("container has no terminal size channel"))]
+    NoTerminalSizeChannel,
+    #[snafu(display("failed to read terminal size"))]
+    GetTerminalSize { source: std::io::Error },
+    #[snafu(display("failed to update terminal size"))]
+    UpdateTerminalSize { source: mpsc::SendError },
+    #[snafu(display("container has no stdin channel"))]
+    NoStdinChannel,
+    #[snafu(display("container has no stdout channel"))]
+    NoStdoutChannel,
+    #[snafu(display("failed to forward stdin to container"))]
+    ForwardStdin { source: std::io::Error },
+    #[snafu(display("failed to forward stdout from container"))]
+    ForwardStdout { source: std::io::Error },
```

Comment:
note: Let's add a few newlines in between variants.

Reply:
1567a8da411f86178862147014e7e1e14ca068a1